bridge/matrix/helpers.go

```diff
@@ -82,22 +83,39 @@ func (b *Bmatrix) getDisplayName(mxid string) string {
 func (b *Bmatrix) cacheDisplayName(mxid string, displayName string) string {
 	now := time.Now()
 
-	// scan to delete old entries, to stop memory usage from becoming too high with old entries
+	// scan to delete old entries, to stop memory usage from becoming too high with old entries.
+	// In addition, we also detect if another user have the same username, and if so, we append their mxids to their usernames to differentiate them.
 	toDelete := []string{}
-	b.RLock()
+	conflict := false
+
+	b.Lock()
 	for k, v := range b.NicknameMap {
+		// to prevent username reuse across matrix servers - or even on the same server, append
+		// the mxid to the username when there is a conflict
+		if v.displayName == displayName {
+			conflict = true
+			// TODO: it would be nice to be able to rename previous messages from this user.
+			// The current behavior is that only users with clashing usernames and *that have spoken since the bridge last started* will get their mxids shown, and I don't know if that's the expected behavior.
+			v.displayName = fmt.Sprintf("%s (%s)", displayName, v.mxid)
```

Comment:
`k` is already the mxid, you don't need to use `v.mxid`
I'd rewrite the for loop to `for mxid,v := range b.NicknameMap...`